src/_pytest/outcomes.py

```diff
@@ -223,12 +230,24 @@ def importorskip(
         # of existing directories with the same name we're trying to
         # import but without a __init__.py file.
         warnings.simplefilter("ignore")
+
+        if exc_type is None:
+            exc_type = ModuleNotFoundError
```

Comment:
Sorry I was not clearer.

I meant that we should only issue the warning if:

1. The user is not passing an explicit `exc_type`.
2. We actually capture `ImportError`.

So something like:


```python
    with warnings.catch_warnings():
        # Make sure to ignore ImportWarnings that might happen because
        # of existing directories with the same name we're trying to
        # import but without a __init__.py file.
        warnings.simplefilter("ignore")

        if exc_type is None:
            exc_type = ImportError
            warn_on_import_error = True
        else:
            warn_on_import_error = False 

        try:
            __import__(modname)
        except exc_type as exc:
            if reason is None:
                reason = f"could not import {modname!r}: {exc}"
            if warn_on_import_error and type(exc) is ImportError:
                warnings.warn(
                    PytestDeprecationWarning(
                        f"pytest.importorskip() caught {exc}, but this will change in a future pytest release to only capture ModuleNotFoundError exceptions by default.\nTo overwrite the future behavior and silence this warning, pass exc_type=ImportError explicitly.",
                    )
                )
            raise Skipped(reason, allow_module_level=True) from None
```

So we will not be changing the behavior right now, only in the future.

Do you agree @The-Compiler?

Reply:
Yep, that sounds like a solid approach. We'll also need tests for the warning and the different argument behaviours.

Reply:
```python
   def importorskip(
    modname: str,
    minversion: Optional[str] = None,
    reason: Optional[str] = None,
    exc_type: Optional[Type[ImportError,ModuleNotFoundError]] = None,
    ) -> Any:
    .
    .
    .
    if exc_type is None:
       #to keep the old behavior
       exc_type = ImportError
       warn_on_import_error = True
    else:
        warn_on_import_error = False
        
    if warn_on_import_error or type(exc) is ImportError:
        warnings.warn()
```
I think the function should be like this, showing a warning if the user doesn't pass any exc_type or when there is ImportError

Reply:
We do not want to issue the warning unless we actually caught `ImportError`.

Most of the time the exception being raised is `ModuleNotFoundError`, which is OK and what users expect. 

The problem is when the library is found, but `ImportError` is raised because of a related problem (for example failure to load a required a shared library). In this case we want to issue a warning now, and in pytest 9.0, let the exception propagate upwards.

The `exc_type` parameter will let users be explicit about what they want to capture.